x-pack/plugin/profiling/src/main/java/org/elasticsearch/xpack/profiling/TransportGetTopNFunctionsAction.java

```diff
@@ -0,0 +1,171 @@
+            StackTrace stackTrace = st.getValue();
+            long samples = response.getStackTraceEvents().getOrDefault(stackTraceId, 0L);
+            builder.addTotalCount(samples);
+
+            int frameCount = stackTrace.frameIds.size();
+            for (int i = 0; i < frameCount; i++) {
+                String frameId = stackTrace.frameIds.get(i);
+                String fileId = stackTrace.fileIds.get(i);
+                Integer frameType = stackTrace.typeIds.get(i);
+                Integer addressOrLine = stackTrace.addressOrLines.get(i);
+                StackFrame stackFrame = response.getStackFrames().getOrDefault(frameId, EMPTY_STACKFRAME);
+                String executable = response.getExecutables().getOrDefault(fileId, "");
+
+                for (Frame frame : stackFrame.frames()) {
+                    String frameGroupId = FrameGroupID.create(fileId, addressOrLine, executable, frame.fileName(), frame.functionName());
+                    if (builder.setCurrentTopNFunction(frameGroupId) == false) {
+                        builder.addTopNFunction(
+                            frameGroupId,
+                            new StackFrameMetadata(
+                                frameId,
+                                fileId,
+                                frameType,
+                                frame.inline(),
+                                addressOrLine,
+                                frame.functionName(),
+                                frame.functionOffset(),
+                                frame.fileName(),
+                                frame.lineNumber(),
+                                executable
+                            )
+                        );
+                    }
+                    if (frameGroupsPerStackTrace.contains(frameGroupId) == false) {
+                        frameGroupsPerStackTrace.add(frameGroupId);
+                        builder.addToCurrentInclusiveCount(samples);
+                    }
+                    if (i == frameCount - 1) {
+                        builder.addToCurrentExclusiveCount(samples);
+                    }
+                }
+            }
+        }
+
+        return builder.build();
+    }
+
+    private static class TopNFunctionsBuilder {
+        private long totalCount = 0;
+        private TopNFunction currentTopNFunction;
+        private final double samplingRate;
+        private final HashMap<String, TopNFunction> topNFunctions;
+
+        TopNFunctionsBuilder(double samplingRate) {
+            this.samplingRate = samplingRate;
+            this.topNFunctions = new HashMap<>();
+        }
+
+        public GetTopNFunctionsResponse build() {
+            List<TopNFunction> functions = new ArrayList<>(topNFunctions.values());
+            Collections.sort(functions, Collections.reverseOrder());
+            long sumSelfCPU = 0;
+            long sumTotalCPU = 0;
+            for (int i = 0; i < functions.size(); i++) {
```

Comment:
Can we use regular iteration here and have a separate `rank` counter that we increase in the body?